### PR TITLE
gntdev: add new ioctls to map dmabuf to existing fd

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
Add new ioctls to allow gntdev to map scatter-gather table on
    top of the existing dmabuf, referenced by file descriptor.
   
    When using dma-buf exporter to create dma-buf with backing storage and
    map it to the grant refs, provided from the domain, we've met a problem,
    that several HW (i.MX8 gpu in our case) do not support external buffer
    and requires backing storage to be created using it's native tools.
    That's why new ioctls were added to be able to pass existing dma-buffer
    fd as input parameter and use it as backing storage to export to refs.
    
    Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

### CODE REVIEW

File: drivers/xen/gntdev-dmabuf.c [hunk 1]
```diff
@@ -443,6 +454,85 @@ dmabuf_exp_alloc_backing_storage(struct gntdev_priv *priv, int dmabuf_flags,
...
+		return ERR_PTR(-ENOMEM);
+
+	if (unlikely(gntdev_test_page_count(count))) {
+		ret = -EINVAL;
+		goto free;
+	}
+
+	dma_buf = dma_buf_get(fd);
+	if (IS_ERR(dma_buf)) {
+		pr_debug("Unable to get dmabuf from fd\n");
```
Comment (lorc) [inline drivers/xen/gntdev-dmabuf.c:526]: `pr_err` will be much better. `dev_err` will be even better

File: drivers/xen/gntdev-dmabuf.c [hunk 2]
```diff
@@ -443,6 +454,85 @@ dmabuf_exp_alloc_backing_storage(struct gntdev_priv *priv, int dmabuf_flags,
...
+	}
+
+	sgt = dma_buf_map_attachment(attach, DMA_BIDIRECTIONAL);
+	if (IS_ERR(sgt)) {
+		ret = PTR_ERR(sgt);
+		goto fail_detach;
+	}
+
+	if (sgt->sgl->offset != data_ofs) {
+		pr_debug("DMA buffer offset %d, user-space expects %d\n",
```
Comment (lorc) [inline drivers/xen/gntdev-dmabuf.c:544]: `pr_warn`?

File: drivers/xen/gntdev-dmabuf.c [hunk 3]
```diff
@@ -443,6 +454,85 @@ dmabuf_exp_alloc_backing_storage(struct gntdev_priv *priv, int dmabuf_flags,
...
+
+	/* Check number of pages that imported buffer has. */
+	if (attach->dmabuf->size < count << PAGE_SHIFT) {
+		pr_debug("DMA buffer has %zu pages, user-space expects %d\n",
+			 attach->dmabuf->size, count);
+		ret = -EINVAL;
+		goto fail_unmap;
+	}
+
+	pr_debug("[gntdev-dmabuf] Map dma-buf fd = %d \n", fd);
```
Comment (lorc) [inline drivers/xen/gntdev-dmabuf.c]: Not needed

File: drivers/xen/gntdev-dmabuf.c [hunk 4]
```diff
@@ -494,6 +584,92 @@ static int dmabuf_exp_from_refs(struct gntdev_priv *priv, int flags,
...
+			gntdev_dmabuf->u.map.map->frames);
+	if (ret != dma_buf->size >> PAGE_SHIFT) {
+		pr_debug("Failed to increase reservation for DMA buffer\n");
+		return ret;
+	}
+
+	xenmem_reservation_va_mapping_update(dma_buf->size >> PAGE_SHIFT,
+			gntdev_dmabuf->u.map.map->pages, gntdev_dmabuf->u.map.map->frames);
+
+	/* Wait on any implicit rendering fences */
```
Comment (lorc) [inline drivers/xen/gntdev-dmabuf.c]: it is not right no mention "rendering" in the generic code.

File: drivers/xen/gntdev-dmabuf.c [hunk 5]
```diff
@@ -494,6 +584,92 @@ static int dmabuf_exp_from_refs(struct gntdev_priv *priv, int flags,
...
+	dma_resv_lock(dma_buf->resv, NULL);
+	ret = dma_resv_wait_timeout_rcu(dma_buf->resv, true, true,
+						  MAX_SCHEDULE_TIMEOUT);
+	if (ret == 0) {
+		pr_debug("dma-buf wasn't clear after timeout. Trying to unmap.\n");
```
Comment (lorc) [inline drivers/xen/gntdev-dmabuf.c]: `err`/`warn`?

File: drivers/xen/gntdev-dmabuf.c [hunk 6]
```diff
@@ -826,6 +1011,47 @@ long gntdev_ioctl_dmabuf_exp_from_refs_v2(struct gntdev_priv *priv, int use_ptem
...
 
+long gntdev_ioctl_dmabuf_map_refs_to_buf(struct gntdev_priv *priv, int use_ptemod,
+					  struct ioctl_gntdev_dmabuf_map_refs_to_buf __user *u)
+{
+	struct ioctl_gntdev_dmabuf_map_refs_to_buf op;
+	u32 *refs;
+	long ret;
+
+	if (use_ptemod) {
+		pr_debug("Cannot provide dma-buf: use_ptemode %d\n",
```
Comment (lorc) [inline drivers/xen/gntdev-dmabuf.c:1073]: `err`/`warn`?

File: drivers/xen/grant-table.c [hunk 7]
```diff
@@ -1220,7 +1220,7 @@ static void __gnttab_unmap_refs_async(struct gntab_unmap_queue_data* item);
 static void gnttab_unmap_work(struct work_struct *work)
 {
 	struct gntab_unmap_queue_data
-		*unmap_data = container_of(work, 
+		*unmap_data = container_of(work,
```
Comment (lorc) [inline drivers/xen/grant-table.c]: spurious change

File: drivers/xen/gntdev-dmabuf.c (hunk 1 above)
Comment (oleksiimoisieiev) [reply drivers/xen/gntdev-dmabuf.c:526]: +

File: drivers/xen/gntdev-dmabuf.c (hunk 2 above)
Comment (oleksiimoisieiev) [reply drivers/xen/gntdev-dmabuf.c:544]: +

File: drivers/xen/gntdev-dmabuf.c (hunk 3 above)
Comment (oleksiimoisieiev) [reply drivers/xen/gntdev-dmabuf.c]: +

File: drivers/xen/gntdev-dmabuf.c (hunk 4 above)
Comment (oleksiimoisieiev) [reply drivers/xen/gntdev-dmabuf.c]: +

File: drivers/xen/gntdev-dmabuf.c (hunk 5 above)
Comment (oleksiimoisieiev) [reply drivers/xen/gntdev-dmabuf.c]: +

File: drivers/xen/gntdev-dmabuf.c (hunk 6 above)
Comment (oleksiimoisieiev) [reply drivers/xen/gntdev-dmabuf.c:1073]: +

File: drivers/xen/grant-table.c (hunk 7 above)
Comment (oleksiimoisieiev) [reply drivers/xen/grant-table.c]: +

**Comment by oleksiimoisieiev** [comment]
Closing PR as obsolete

**Pull request closed without merging**